### PR TITLE
[Segment Replication] Add check to cancel ongoing replication with old primary on onNewCheckpoint on replica

### PR DESCRIPTION
**Pull request by dreamer-89**
Signed-off-by: Suraj Singh <surajrider@gmail.com>

### Description
During failover, when a new primary is chosen, replica should close on-going replication with previous primary. Continuing to copy files from two different primaries can be catastrophic and lead to multiple failures e.g. shard failure. Though logically two primary can't co-exist but due to some random bug, an older primary node can continue to hold onto replica for file transfer. This change is also needed:
1. To clean local store and cleanup temporary files already copied. 
2. Interrupt the existing threads involved in replication. 

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/4205

### Related
https://github.com/opensearch-project/OpenSearch/pull/4225

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dreamer-89** [comment]
Last gradle check failure due to heap space
```
* What went wrong:
Execution failed for task ':plugins:analysis-smartcn:forbiddenApisTest'.
> Java heap space
```

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 1]
```diff
@@ -151,14 +152,26 @@ public synchronized void onNewCheckpoint(final ReplicationCheckpoint receivedChe
...
-                )
-            );
-            return;
+        Optional<SegmentReplicationTarget> ongoingReplicationTarget = onGoingReplications.getOngoingReplicationTarget(
+            replicaShard.shardId()
+        );
+        if (ongoingReplicationTarget.isPresent()) {
+            final SegmentReplicationTarget target = ongoingReplicationTarget.get();
+            if (target.getCheckpoint().getPrimaryTerm() < receivedCheckpoint.getPrimaryTerm()) {
+                logger.info(
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:161]: nit - log level here can also be trace.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 2]
```diff
@@ -151,14 +152,26 @@ public synchronized void onNewCheckpoint(final ReplicationCheckpoint receivedChe
...
+                    "Cancelling ongoing replication from old primary with primary term {}",
+                    target.getCheckpoint().getPrimaryTerm()
+                );
+                target.cancel("Cancelling stuck target after new primary");
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:165]: This should be invoking cancel through onGoingSegmentReplications cancel method so that it is removed from the collection.

File: server/src/main/java/org/opensearch/indices/replication/common/ReplicationCollection.java [hunk 3]
```diff
@@ -236,13 +237,13 @@ public boolean cancelForShard(ShardId shardId, String reason) {
...
      *
-     * @param shardId      shardId for which to check if replicating
-     * @return true if shard is currently replicating
+     * @param shardId      shardId
+     * @return Optional ReplicationTarget for input shardId
      */
-    public boolean isShardReplicating(ShardId shardId) {
-        return onGoingTargetEvents.values().stream().anyMatch(t -> t.indexShard.shardId().equals(shardId));
+    public Optional<T> getOngoingReplicationTarget(ShardId shardId) {
+        return onGoingTargetEvents.values().stream().filter(t -> t.indexShard.shardId().equals(shardId)).findFirst();
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/replication/common/ReplicationCollection.java:246]: Thinking we should add an assert here that there is only a single entry?

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java (hunk 1 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:161]: Done

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java (hunk 2 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:165]: Updated

File: server/src/main/java/org/opensearch/indices/replication/common/ReplicationCollection.java (hunk 3 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/indices/replication/common/ReplicationCollection.java]: I think we should also assert while adding starting new replications. Existing asserts is on target id which is AtomicLong and will always be unique. 
```
T existingTarget = onGoingTargetEvents.putIfAbsent(target.getId(), target);
assert existingTarget == null : "found two Target instances with the same id";
```

Comment (dreamer-89) [reply server/src/main/java/org/opensearch/indices/replication/common/ReplicationCollection.java]: I will take this in a follow up PR. Let me know if you think otherwise.

**Review by mch2 — APPROVED** [review]
lgtm, thanks @dreamer-89

**Comment by dreamer-89** [comment]
```
org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > classMethod FAILED
    com.carrotsearch.randomizedtesting.ResourceDisposalError: Resource in scope SUITE failed to close. Resource was registered from thread Thread[id=117, name=TEST-SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication-seed#[CB89A68B01537D55], state=RUNNABLE, group=TGRP-SegmentReplicationTargetServiceTests], registration stack trace below.
        at __randomizedtesting.SeedInfo.seed([CB89A68B01537D55]:0)
        at java.****@17.0.4/java.lang.Thread.getStackTrace(Thread.java:1610)
        at app//com.carrotsearch.randomizedtesting.RandomizedContext.closeAtEnd(RandomizedContext.java:157)
        at app//org.apache.lucene.tests.util.LuceneTestCase.closeAfterSuite(LuceneTestCase.java:804)
        at app//org.apache.lucene.tests.util.LuceneTestCase.wrapDirectory(LuceneTestCase.java:1458)
        at app//org.apache.lucene.tests.util.LuceneTestCase.newFSDirectory(LuceneTestCase.java:1409)
        at app//org.apache.lucene.tests.util.LuceneTestCase.newFSDirectory(LuceneTestCase.java:1387)
        at app//org.apache.lucene.tests.util.LuceneTestCase.newFSDirectory(LuceneTestCase.java:1374)
        at app//org.opensearch.index.shard.IndexShardTestCase.createStore(IndexShardTestCase.java:234)
        at app//org.opensearch.index.shard.IndexShardTestCase.lambda$newShard$4(IndexShardTestCase.java:523)
        at app//org.opensearch.index.shard.IndexShardTestCase.newShard(IndexShardTestCase.java:525)
        at app//org.opensearch.index.shard.IndexShardTestCase.newShard(IndexShardTestCase.java:475)
        at app//org.opensearch.index.shard.IndexShardTestCase.newShard(IndexShardTestCase.java:433)
        at app//org.opensearch.index.shard.IndexShardTestCase.newShard(IndexShardTestCase.java:318)
        at app//org.opensearch.index.shard.IndexShardTestCase.newShard(IndexShardTestCase.java:280)
        at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.setUp(SegmentReplicationTargetServiceTests.java:63)

        Caused by:
        java.lang.AssertionError: Directory not closed: MockDirectoryWrapper(NIOFSDirectory@/var/jenkins/workspace/gradle-check/search/server/build/testrun/test/temp/org.opensearch.indices.replication.SegmentReplicationTargetServiceTests_CB89A68B01537D55-001/tempDir-022/indices/_na_/0/index lockFactory=org.apache.lucene.store.NativeFSLockFactory@2fb70b2d)
            at org.junit.Assert.fail(Assert.java:89)
            at org.apache.lucene.tests.util.CloseableDirectory.close(CloseableDirectory.java:43)
            at com.carrotsearch.randomizedtesting.RandomizedRunner$4$1.apply(RandomizedRunner.java:784)
            at com.carrotsearch.randomizedtesting.RandomizedRunner$4$1.apply(RandomizedRunner.java:781)
            at com.carrotsearch.randomizedtesting.RandomizedContext.closeResources(RandomizedContext.java:225)
            at com.carrotsearch.randomizedtesting.RandomizedRunner$4.afterAlways(RandomizedRunner.java:797)
            at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:43)
            at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
            at java.****/java.lang.Thread.run(Thread.java:833)

```

**Comment by dreamer-89** [comment]
```
**1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':distribution:bwc:bugfix:buildBwcLinuxTar'.
> Building 2.2.1 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/bugfix/build/bwc/checkout-2.2/distribution/archives/linux-tar/build/distributions/opensearch-min-2.2.1-SNAPSHOT-linux-x64.tar.gz

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.**

```

**Comment by dreamer-89** [comment]
```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication" -Dtests.seed=A5B34090509BB67E -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fr-LU -Dtests.timezone=US/Arizona -Druntime.java=17

org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testOnNewCheckpointFromNewPrimaryCancelOngoingReplication FAILED
    org.mockito.exceptions.verification.TooManyActualInvocations: 
    segmentReplicationTargetService.startReplication(
        ReplicationCheckpoint{shardId=[index][0], primaryTerm=37, segmentsGen=2, seqNo=-1, version=6},
        org.opensearch.index.shard.IndexShard@429facb4,
        <any>
    );
    Wanted 1 time:
    -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
    But was 2 times:
    -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
    -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
        at __randomizedtesting.SeedInfo.seed([A5B34090509BB67E:7C8E89BED0C8154F]:0)
        at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
```

**Comment by mch2** [comment]
> This is not reproducible locally even after multiple retries. Refiring!
> 
> ```
> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication" -Dtests.seed=A5B34090509BB67E -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fr-LU -Dtests.timezone=US/Arizona -Druntime.java=17
> 
> org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testOnNewCheckpointFromNewPrimaryCancelOngoingReplication FAILED
>     org.mockito.exceptions.verification.TooManyActualInvocations: 
>     segmentReplicationTargetService.startReplication(
>         ReplicationCheckpoint{shardId=[index][0], primaryTerm=37, segmentsGen=2, seqNo=-1, version=6},
>         org.opensearch.index.shard.IndexShard@429facb4,
>         <any>
>     );
>     Wanted 1 time:
>     -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
>     But was 2 times:
>     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
>     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
>         at __randomizedtesting.SeedInfo.seed([A5B34090509BB67E:7C8E89BED0C8154F]:0)
>         at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
> ```

Hmm not sure whats going on here but this test is a bit hard to follow.  I'm wondering if this case is better tested without all the spy/mocks in `SegmentReplicationIndexShardTests` where we can stub out the replication source and simulate a new request arriving similar to [this](https://github.com/opensearch-project/OpenSearch/blob/19d1a2b027fef8b981560969bf428476d700bd07/server/src/test/java/org/opensearch/index/shard/SegmentReplicationIndexShardTests.java#L268) test.

**Comment by dreamer-89** [comment]
Flaky test, refiring!
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.cluster.routing.allocation.decider.DiskThresholdDeciderIT.testRestoreSnapshotAllocationDoesNotExceedWatermark" -Dtests.seed=AE07C93EFEF373ED -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sv-SE -Dtests.timezone=Etc/GMT-9 -Druntime.java=17

org.opensearch.cluster.routing.allocation.decider.DiskThresholdDeciderIT > testRestoreSnapshotAllocationDoesNotExceedWatermark FAILED
    java.lang.AssertionError: 
    Expected: an empty collection
         but: <[[vynmviietw][0], node[yA2a8fg5TC-YBclcuDS8jQ], [P], s[STARTED], a[id=51NXkDtVRXOTEdfxctsFHw], [vynmviietw][4], node[yA2a8fg5TC-YBclcuDS8jQ], [P], s[STARTED], a[id=Vux_vctAS2mn_IjAYd073w]]>
        at __randomizedtesting.SeedInfo.seed([AE07C93EFEF373ED:BCBF6F5F954A6C55]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:964)
        at org.junit.Assert.assertThat(Assert.java:930)
        at org.opensearch.cluster.routing.allocation.decider.DiskThresholdDeciderIT.lambda$testRestoreSnapshotAllocationDoesNotExceedWatermark$2(DiskThresholdDeciderIT.java:260)
        at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1049)
        at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1022)
        at org.opensearch.cluster.routing.allocation.decider.DiskThresholdDeciderIT.testRestoreSnapshotAllocationDoesNotExceedWatermark(DiskThresholdDeciderIT.java:260)

```

**Comment by dreamer-89** [comment]
> > This is not reproducible locally even after multiple retries. Refiring!
> > ```
> > REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication" -Dtests.seed=A5B34090509BB67E -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fr-LU -Dtests.timezone=US/Arizona -Druntime.java=17
> > 
> > org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testOnNewCheckpointFromNewPrimaryCancelOngoingReplication FAILED
> >     org.mockito.exceptions.verification.TooManyActualInvocations: 
> >     segmentReplicationTargetService.startReplication(
> >         ReplicationCheckpoint{shardId=[index][0], primaryTerm=37, segmentsGen=2, seqNo=-1, version=6},
> >         org.opensearch.index.shard.IndexShard@429facb4,
> >         <any>
> >     );
> >     Wanted 1 time:
> >     -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
> >     But was 2 times:
> >     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
> >     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
> >         at __randomizedtesting.SeedInfo.seed([A5B34090509BB67E:7C8E89BED0C8154F]:0)
> >         at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
> > ```
> 
> Hmm not sure whats going on here unless its replaying the latestReceivedCheckpoint so the checkpoint is processed twice? I'm wondering if this case is better tested without all the spy/mocks in `SegmentReplicationIndexShardTests` where we can stub out the replication source and simulate a new request arriving similar to [this](https://github.com/opensearch-project/OpenSearch/blob/19d1a2b027fef8b981560969bf428476d700bd07/server/src/test/java/org/opensearch/index/shard/SegmentReplicationIndexShardTests.java#L268) test.

Thanks @mch2, I rebased against main and this test is failing anymore (without rebase it was failing 50% times). I think this test fairly stable now, can I take up `SegmentReplicationIndexShardTests` test in separate PR ?


Screen Shot 2022-09-02 at 11 03 19 AM

**Comment by dreamer-89** [comment]
> > > This is not reproducible locally even after multiple retries. Refiring!
> > > ```
> > > REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication" -Dtests.seed=A5B34090509BB67E -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fr-LU -Dtests.timezone=US/Arizona -Druntime.java=17
> > > 
> > > org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testOnNewCheckpointFromNewPrimaryCancelOngoingReplication FAILED
> > >     org.mockito.exceptions.verification.TooManyActualInvocations: 
> > >     segmentReplicationTargetService.startReplication(
> > >         ReplicationCheckpoint{shardId=[index][0], primaryTerm=37, segmentsGen=2, seqNo=-1, version=6},
> > >         org.opensearch.index.shard.IndexShard@429facb4,
> > >         <any>
> > >     );
> > >     Wanted 1 time:
> > >     -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
> > >     But was 2 times:
> > >     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
> > >     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
> > >         at __randomizedtesting.SeedInfo.seed([A5B34090509BB67E:7C8E89BED0C8154F]:0)
> > >         at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
> > > ```
> > 
> > 
> > Hmm not sure whats going on here unless its replaying the latestReceivedCheckpoint so the checkpoint is processed twice? I'm wondering if this case is better tested without all the spy/mocks in `SegmentReplicationIndexShardTests` where we can stub out the replication source and simulate a new request arriving similar to [this](https://github.com/opensearch-project/OpenSearch/blob/19d1a2b027fef8b981560969bf428476d700bd07/server/src/test/java/org/opensearch/index/shard/SegmentReplicationIndexShardTests.java#L268) test.
> 
> Thanks @mch2, I rebased against main and this test is failing anymore (without rebase it was failing 50% times). I think this test fairly stable now, can I take up `SegmentReplicationIndexShardTests` test in separate PR ?
> 
> Screen Shot 2022-09-02 at 11 05 54 AM

This test is still flaky fails ~1% times caused gradle check failures on CI.

**Comment by dreamer-89** [comment]
> > > > This is not reproducible locally even after multiple retries. Refiring!
> > > > ```
> > > > REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication" -Dtests.seed=A5B34090509BB67E -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fr-LU -Dtests.timezone=US/Arizona -Druntime.java=17
> > > > 
> > > > org.opensearch.indices.replication.SegmentReplicationTargetServiceTests > testOnNewCheckpointFromNewPrimaryCancelOngoingReplication FAILED
> > > >     org.mockito.exceptions.verification.TooManyActualInvocations: 
> > > >     segmentReplicationTargetService.startReplication(
> > > >         ReplicationCheckpoint{shardId=[index][0], primaryTerm=37, segmentsGen=2, seqNo=-1, version=6},
> > > >         org.opensearch.index.shard.IndexShard@429facb4,
> > > >         <any>
> > > >     );
> > > >     Wanted 1 time:
> > > >     -> at org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
> > > >     But was 2 times:
> > > >     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
> > > >     -> at org.opensearch.indices.replication.SegmentReplicationTargetService.onNewCheckpoint(SegmentReplicationTargetService.java:174)
> > > >         at __randomizedtesting.SeedInfo.seed([A5B34090509BB67E:7C8E89BED0C8154F]:0)
> > > >         at app//org.opensearch.indices.replication.SegmentReplicationTargetServiceTests.testOnNewCheckpointFromNewPrimaryCancelOngoingReplication(SegmentReplicationTargetServiceTests.java:235)
> > > > ```
> > > 
> > > 
> > > Hmm not sure whats going on here unless its replaying the latestReceivedCheckpoint so the checkpoint is processed twice? I'm wondering if this case is better tested without all the spy/mocks in `SegmentReplicationIndexShardTests` where we can stub out the replication source and simulate a new request arriving similar to [this](https://github.com/opensearch-project/OpenSearch/blob/19d1a2b027fef8b981560969bf428476d700bd07/server/src/test/java/org/opensearch/index/shard/SegmentReplicationIndexShardTests.java#L268) test.
> > 
> > 
> > Thanks @mch2, I rebased against main and this test is failing anymore (without rebase it was failing 50% times). I think this test fairly stable now, can I take up `SegmentReplicationIndexShardTests` test in separate PR ?
> > Screen Shot 2022-09-02 at 11 05 54 AM
> 
> This test is still flaky fails ~1% times caused gradle check failures on CI.

Raised https://github.com/opensearch-project/OpenSearch/pull/4414
